allowlists/actions/getAllowListRecordsForAddressByClaimedWithMetadata.tsx

```diff
@@ -0,0 +1,106 @@
+import "server-only";
+
+import { ResultOf, graphql, readFragment } from "@/lib/graphql";
+
+import request from "graphql-request";
+import { HYPERCERTS_API_URL_GRAPH } from "@/configs/hypercerts";
+import { getHypercertMetadata } from "@/hypercerts/actions/getHypercertMetadata";
+import { UnclaimedFraction } from "@/components/profile/unclaimed-hypercerts-list";
+
+export const AllowListRecordFragment = graphql(`
+  fragment AllowListRecordFragment on AllowlistRecord {
+    id
+    hypercert_id
+    token_id
+    root
+    leaf
+    entry
+    user_address
+    claimed
+    proof
+    units
+    total_units
+  }
+`);
+export type AllowListRecord = ResultOf<typeof AllowListRecordFragment>;
+
+const query = graphql(
+  `
+    query allowlistRecords($address: String, $claimed: Boolean) {
+      allowlistRecords(
+        where: { user_address: { eq: $address }, claimed: { eq: $claimed } }
+      ) {
+        count
+        data {
+          ...AllowListRecordFragment
+        }
+      }
+    }
+  `,
+  [AllowListRecordFragment],
+);
+
+const requestMap = new Map<string, Promise<any>>();
+
+async function getMetadataWithDeduping(hypercertId: string) {
+  if (requestMap.has(hypercertId)) {
+    return requestMap.get(hypercertId);
+  }
+
+  const requestPromise = getHypercertMetadata(hypercertId);
+  requestMap.set(hypercertId, requestPromise);
+
+  try {
+    const result = await requestPromise;
+    return result;
+  } finally {
+    requestMap.delete(hypercertId);
```

Comment:
`try {} finally {}`, not seeing the point of the syntax without a catch. Wondering if we should delete here, as it seems more logical to clear the entire map once all have resolved, to prevent doing the request twice after all.

Reply:
updated it to return the promise. the requestMap can live for the lifecycle of the method. Pushed the changes